### ISSUE TITLE
[accessibility] VoiceOver on toggleable Material widgets on iOS

### ISSUE THREAD

**Issue by TahaTesser** [flutter/flutter#121636]
Internal: b/275699766

While fixing https://github.com/flutter/flutter/issues/120650 and investigating https://github.com/flutter/flutter/issues/74963, I noticed that regular `ListTile` doesn't announce any action and other toggleable material widgets with voice-over doesn't sound right compared to Android. 

Since these toggleable widgets are integrable parts of list tile-dependent widgets such as `CheckboxListTile`, `SwitchListTile`, `RadioListTile`, and `ExpansionTile`. I wanna make sure the hints are consistent on iOS for all these widgets.

Here I created an example that showcases toggleable Material widgets and their list tile variants.

<details> 
<summary>code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      theme: ThemeData(useMaterial3: true),
      home: const Example(),
    );
  }
}

class Example extends StatefulWidget {
  const Example({super.key});

  @override
  State<Example> createState() => _ExampleState();
}

class _ExampleState extends State<Example> {
  bool toggle = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: SingleChildScrollView(
          child: Column(
            children: <Widget>[
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: <Widget>[
                  Column(
                    children: <Widget>[
                      const DescriptionText(text: 'Switch'),
                      Switch(
                        value: toggle,
                        onChanged: (bool value) {
                          setState(() {
                            toggle = !toggle;
                          });
                        },
                      ),
                    ],
                  ),
                  Column(
                    children: <Widget>[
                      const DescriptionText(text: 'Checkbox'),
                      Checkbox(
                        value: toggle,
                        onChanged: (bool? value) {
                          setState(() {
                            toggle = !toggle;
                          });
                        },
                      ),
                    ],
                  ),
                  Column(
                    children: <Widget>[
                      const DescriptionText(text: 'Radio'),
                      Radio<int>(
                        value: toggle ? 1 : 2,
                        groupValue: 1,
                        toggleable: true,
                        onChanged: (int? value) {
                          setState(() {
                            toggle = !toggle;
                          });
                        },
                      ),
                    ],
                  )
                ],
              ),
              const Divider(),
              const DescriptionText(text: 'ListTile'),
              ListTile(
                title: const Text('Headline'),
                trailing: const Icon(Icons.favorite),
                selected: toggle,
                onTap: () {
                  setState(() {
                    toggle = !toggle;
                  });
                },
              ),
              const Divider(),
              const DescriptionText(text: 'SwitchListTile'),
              SwitchListTile(
                title: const Text('Headline'),
                value: toggle,
                onChanged: (bool value) {
                  setState(() {
                    toggle = !toggle;
                  });
                },
              ),
              const Divider(),
              const DescriptionText(text: 'CheckboxListTile'),
              CheckboxListTile(
                title: const Text('Headline'),
                value: toggle,
                onChanged: (bool? value) {
                  setState(() {
                    toggle = !toggle;
                  });
                },
              ),
              const Divider(),
              const DescriptionText(text: 'RadioListTile'),
              RadioListTile<int>(
                title: const Text('Headline'),
                value: toggle ? 1 : 2,
                groupValue: 1,
                toggleable: true,
                onChanged: (int? value) {
                  setState(() {
                    toggle = !toggle;
                  });
                },
              ),
              const Divider(),
              const DescriptionText(text: 'ExpansionTile'),
              const ExpansionTile(
                title: Text('Headline'),
                children: <Widget>[FlutterLogo(size: 50)],
              ),
              const Divider(),
            ],
          ),
        ),
      ),
    );
  }
}

class DescriptionText extends StatelessWidget {
  const DescriptionText({super.key, required this.text});

  final String text;

  @override
  Widget build(BuildContext context) {
    return ExcludeSemantics(child: Text(text));
  }
}

``` 
	
</details>

### Preview
https://user-images.githubusercontent.com/48603081/221981458-bc13a451-a351-4c08-a859-46b3327071cc.MP4

https://user-images.githubusercontent.com/48603081/221981497-6cc2ee05-d4c7-4d05-ac78-3d9128333b68.MP4

### Observations

- All toggleable widgets are announced as 'switch button" on iOS.
- No double tap to expand announcement for expansion tile https://github.com/flutter/flutter/issues/74963
- The radio button is never announced due to https://github.com/flutter/flutter/issues/80182

Fix voiceover issues for the following widgets:

- [ ] [Switch](https://api.flutter.dev/flutter/material/Switch-class.html)
- [ ] [Checkbox](https://api.flutter.dev/flutter/material/Checkbox-class.html)
- [ ] [Radio](https://api.flutter.dev/flutter/material/Radio-class.html)
- [ ] [CheckboxListTile](https://api.flutter.dev/flutter/material/CheckboxListTile-class.html)
- [ ] [SwitchListTile](https://api.flutter.dev/flutter/material/SwitchListTile-class.html)
- [ ] [RadioListTile](https://api.flutter.dev/flutter/material/RadioListTile-class.html)
- [ ] [ExpansionTile](https://api.flutter.dev/flutter/material/ExpansionTile-class.html)

**Comment by TahaTesser** [comment]
cc: @chunhtai 
I investigated earlier today. I think this is worth addressing along ListTile and ExpansionTile for semantics for iOS as you pointed label doesn't; work on iOS.

**Comment by chunhtai** [comment]
I assigned this issue to you @TahaTesser since you are working on the expand and collapse pr, are you able to work on other control as well? Feel free to unassign and assign it back to me if this is not in the scope of your work

**Comment by TahaTesser** [comment]
> I assigned this issue to you @TahaTesser since you are working on the expand and collapse pr, are you able to work on other control as well? Feel free to unassign and assign it back to me if this is not in the scope of your work

Sure can :)

**Comment by TahaTesser** [comment]
@chunhtai 

After looking into this closely, it seems like `Semantics.toggled` & `Semantics.checked`  are all flagged as switched on the engine side for iOS.  
I'm not very familiar with objective-c with fix this myself, I found these [lines](https://github.com/flutter/engine/blob/5f517d646231b8f63e291f836bc5ec5f273bc472/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm#L270-L272) in the engine that look similar to the behavior.

Feel free to take it or assign someone else

**Comment by jiahaog** [comment]
Marking as P2 because b/275699766 is marked as P1

**Comment by mariamhas** [comment]
updating based on timeline needed for fix

**Comment by chunhtai** [comment]
@TahaTesser

@mariamhas Can you also help me reach out to the customers?

I am not entirely what is the expected behavior

ListTile, CheckboxListTile, SwitchListTile, RadioListTile, and ExpansionTile. For all these right now it would announce
"<label>, swtich button, on/off, double tap to toggle settings"

I looked at the counter part in swift UI
The only difference is the ExpansionTile where it would be announced as
"label, collpased, double tap to expand"

is the ExpansionTitle the only issue here?

**Comment by TahaTesser** [comment]
> ListTile, CheckboxListTile, SwitchListTile, RadioListTile, and ExpansionTile. For all these right now it would announce
> ", swtich button, on/off, double tap to toggle settings"

`ExpansionTile` contains `ListTile`, which doesn't announce any action, not even "double tap to toggle settings" .

My earlier fix was just for `ExpansionTile`, which overrides the `label` but a much better solution would be to address for `ListTile` and then append `collapsed` or `expanded` state and `double tap to expand" instead of `double tap toggle settings`.  

Also, Switch has announced the switch, the checkbox is also announced switch, radio is also announced as a switch.  Fixing these will also require adding labels since hints don't work on iOS, which should be properly merged with list tile semantics without conflict or duplicate values for `SwitchListTile`, `RadioListTile`, `CheckboxListTile`. 

The problem for me is this is much more complex to address from` InkWell` -> ListTile -> ExpansionTile without making sure it doesn't break all these widgets or tests. `InkWell` semantics don't bother 

> is the ExpansionTitle the only issue here?

This could be addressed directly but if `ListTile` needs the same fix later, we get double announcements since the `ExpansionTile` fix doesn't account for `ListTile` semantics.

Something like:
`label, expanded, double to expand for more details, double to toggle settings` when merged semantics. 

I'm hoping you can better address this than I can.

**Comment by chunhtai** [comment]
SwitchListTile, RadioListTile, and CheckboxListTile are all Toggle in swiftUI with different style. They will all be announced as switch button, so I think I will leave them as is.

Our ListTile is too generic to have any opinion on the its semantics property. It can built to be a button, or built to be a toggleable tile. I think I would leave it not opinionated.

then what's left is to make ExpansionTile to be treated as non switch button but was able to announce correct message.
